File: src/napakalaki/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package napakalaki;

import java.util.ArrayList;
import java.util.Random;

public class Napakalaki {
    private Player currentPlayer;
    private ArrayList<Player> players;
    private CardDealer dealer = CardDealer.getInstance();
    private Monster currentMonster;
    private CombatResult result;
            
    private Napakalaki() {
    }
    
    public static Napakalaki getInstance() {
        return NapakalakiHolder.INSTANCE;
    }
    
    private static class NapakalakiHolder {

        private static final Napakalaki INSTANCE = new Napakalaki();
    }
    
    private void initPlayers(ArrayList<String> names){
        players = new ArrayList<>();
        for (String name : names) {
            players.add(new Player(name));
        }
    }
    
    private Player nextPlayer(){ 
        if (currentPlayer == null){
            Random rnd = new Random();
            currentPlayer = players.get(rnd.nextInt(players.size()));
        }else{
            if (currentPlayer == players.get(players.size()-1)){
                currentPlayer = players.get(0);
            }else{
                currentPlayer = players.get(players.indexOf(currentPlayer)+1);
            }
        }
        return currentPlayer;
    }
    
    private boolean nextTurnAllowed(){
        return currentPlayer == null || currentPlayer.validState();
    }
    
    public CombatResult developCombat(){
        CombatResult combat;
        combat = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        return combat;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for (Treasure treasure : treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        for (Treasure treasure : treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void makeTreasuresVisibles(ArrayList<Treasure> treasures){
        for (Treasure t: treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public boolean buyLevels(ArrayList<Treasure> visible, ArrayList<Treasure> hidden){
        boolean canI;
        canI = currentPlayer.buyLevels(visible, hidden);
        return canI;
    }
    
    public void initGame(ArrayList<String> players){
        initPlayers(players);
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK;
        stateOK = nextTurnAllowed();
        if (stateOK) {
            currentMonster = dealer.nextMonsters();
            currentPlayer = nextPlayer();
            boolean dead = currentPlayer.isDead();
            if (dead){
                currentPlayer.initTreasures();
            }
        }
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        return (result == CombatResult.WinAndWinGame);
    }
}

File: P5Ruby/NapakalakiRuby/NapakalakiRuby/lib/napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

class Napakalaki
   attr_accesor :current_player, :players, :dealer, :current_monster, :result
  include Singleton
  def initialize
    
  end
  private
  def init_players(names)
    @players = Array.new
    
    names.each { |i|   
       @players << player.new(i)
     
    }
    
  end
  def next_player
    
    if(@current_player == nil)
      @current_player = @players.get(rand(@players.size))
    else
      
      if(@current_player  == @players.get(@players.size))
        @current_player = @players.get(0);
      else
          
       @current_player = @players.get(@players.index(@current_player)+1)  
        
      end
       
    end 
      
  return @current_player
    
  end
  def next_turn_allowed
    return @current_player.valid_state
  end
  public
  def develop_combat
    
    monster = @current_monster
    
    @current_player.combat(monster)
    
      
      
      
      
      
      
      
      
      
      
    end
    
    
    
  end
  def discard_visible_treasures(treasures)
    
    
    treasures.each do |tr|
      
       @current_player.discard_visible_treasure(tr)
       @current_player.give_treasure_back(tr)
      
    end
   
   
    
  end
  def discard_hidden_treasures(treasures)
    
    treasures.each do |tr|
      
       @current_player.discard_hidden_treasure(tr)
       @current_player.give_treasure_back(tr)
      
    end
   
  end
  def make_treasures_visible(treasures)
    
  end
  def buy_levels(visible, hidden)
    
  end
  def init_game(players)
    
    @dealer.init_players(players)
    @dealer.init_cards 
    next_turn
    
    
   
  end
  def get_current_player
    return @current_player
  end
  def get_current_monster
    return @current_monster
  end
  def next_turn
    
    stateOK = next_turn_allowed
    
    if stateOK
      
      @current_monster = dealer.next_monster
      @current_player = next_player
      dead = @current_player.is_dead
      
      if dead
        
        @current_player.init_treasures
        
        
      end
      
      
    end
          
    
  end
  
  
  def end_of_game(result)
    return(result == CombatResult::WINANDWINGAME)
  end
end
